pint/compat/__init__.py

```diff
@@ -85,6 +106,23 @@ def _to_magnitude(value, force_ndarray=False):
             return np.asarray(value)
         return value
 
+    def _test_array_function_protocol():
+        # Test if the __array_function__ protocol is enabled
+        try:
+            class FakeArray:
+                def __array_function__(self, *args, **kwargs):
+                    return
+
+            np.concatenate([FakeArray()])
+            return True
+        except ValueError:
+            return False
+
+    HAS_NUMPY_ARRAY_FUNCTION = _test_array_function_protocol()
+
+    if HAS_NUMPY_ARRAY_FUNCTION:
+        warnings.warn(_msg, BehaviorChangeWarning)
```

Comment:
This is going to annoy a lot of people who don't use pint with numpy but just happen to have numpy 1.17 installed in the same environment.
Could we issue the warning upon the first call to ``__array_function__``? Or upon the creation of the first Quantity with numpy data?

Reply:
Sounds good! I agree that would be annoying. Which of the two (warn on first `__array_function__` or warn on first creation) would make the most sense here? 

Also, would the following (for example in the warn-on-first-creation case) be a good replacement for the suggested way to hide the warning:

```python
import numpy as np
import warnings
from pint import Quantity

with warnings.catch_warnings():
    q = Quantity([1, 2, 3, 4], 'm')
```

or would something else be better?

Reply:
did we resolve this, yet? If not, I'd vote for raising on first creation since that would make sure people know that what they just created behaves different from before.

To silence it, I think the sample above works. However, I wouldn't assign a name to it (since that can make linters like `flake8` complain about a name being assigned but not used). We could also shorten it to `Quantity([], 'm')` or use `Quantity(np.array([]), "m")`?

Reply:
Since it seems like we have enough of a consensus, I've just pushed a commit to change the warning to on-first-creation.